backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/snapshots/CreateSnapshotForVmCommand.java

```diff
@@ -441,11 +444,27 @@ protected void endVmCommand() {
                 log.warn("No snapshot was created for VM '{}' which is in LOCKED status", getVmId());
             } else {
                 Snapshot createdSnapshot = snapshotDao.get(getParameters().getCreatedSnapshotId());
-                revertToActiveSnapshot(createdSnapshot.getId());
-                // If the removed snapshot contained memory, remove the memory volumes
-                // Note that the memory volumes might not have been created
-                if (snapshotWithMemory(createdSnapshot)) {
-                    removeMemoryVolumesOfSnapshot(createdSnapshot);
+                boolean snapshotUsed = false;
+                for (DiskImage disk: getDisksList()) {
+                    DiskImage snapshotImage = diskImageDao.getDiskSnapshotForVmSnapshot(disk.getId(),
+                            createdSnapshot.getId());
+                    if (snapshotImage != null && imagesHandler.isSnapshotUsed(getVm(), snapshotImage)) {
+                        snapshotUsed = true;
+                        break;
+                    }
+                }
+
+                if (snapshotUsed) {
+                    log.warn("Snapshot '{}' appears to be in use despite failure, cleanup may be required",
+                            createdSnapshot.getId());
+                    unlockSnapshot(createdSnapshot.getId());
```

Comment:
When setting the snapshot to `OK` status here. Will it be usable to every operation afterwards? Is it possible if the snapshot flow failed? Will it make sense to set it to another status, which the snapshot will appear but won't be usable?

Reply:
The snapshot flow did fail when we got here but we managed to create the overlay volume, so in theory it should be fine and removable, but I am not sure about a situation where multiple disks are included in the snapshot and some overlays are created while others aren't

And now I also noticed I am checking if the incorrect volumes are used, and I need to check the volumes in the new Active VM snapshot

Reply:
> The snapshot flow did fail when we got here but we managed to create the overlay volume, so in theory it should be fine and removable, but I am not sure about a situation where multiple disks are included in the snapshot and some overlays are created while others aren't

I think Liran meant to something different - so we will end up with the snapshot being in OK state while including some volumes that contain the state at that point in time and some volumes that contain the up-to-date data (as they server as active volumes) and so if the user tries to restore that snapshot or clone a vm from that snapshot, the created entity won't have the whole state of the vm at the point where the snapshot was taken

we don't have an 'illegal' state to snapshots.. do we have any indication that the snapshot might be corrupted/inconsistent? if we remove the snapshot from the database then I guess we won't be able to remove the extra volumes..

> And now I also noticed I am checking if the incorrect volumes are used, and I need to check the volumes in the new Active VM snapshot

yeah, I had a comment on this that I removed now :)

Reply:
>I think Liran meant to something different - so we will end up with the snapshot being in OK state while including some volumes that contain the state at that point in time and some volumes that contain the up-to-date data (as they server as active volumes) and so if the user tries to restore that snapshot or clone a vm from that snapshot, the created entity won't have the whole state of the vm at the point where the snapshot was taken

yes, I mean that what is going to be "illegal" is the existing volume, but is there a reason where it would not be usable? I think the new active image can probably corrupted by this operation but I'm not sure about the parent volume

>we don't have an 'illegal' state to snapshots.. do we have any indication that the snapshot might be corrupted/inconsistent? if we remove the snapshot from the database then I guess we won't be able to remove the extra volumes..

Perhaps we could mark the Active VM images as illegal in the engine and force the user to do something,
not unlocking the snapshot will just leave it locked which I think is less desirable

Reply:
We definitely  shouldn't keep the snapshot locked, no question about that, but maybe we can find a better state instead of having the original snapshot in OK state.

I don't see why a volume, either a snapshot volume or a new active, would be corrupted

How about modifying the snapshot instead of keeping it as it? Let's say that the snapshot (pivot) worked for disk A and didn't work for disk B so in that case we can turn the snapshot to a partial snapshot that contains only disk A and remove the supposed to be new-active volume for disk B.

Another case may be a snapshot with memory for a single-disk VM - if the pivot succeeded but the snapshot operation failed to dump the memory, we'll turn the snapshot to be a disk-only snapshot

That way we shouldn't end up with a snapshot that points to volumes that are still active and thus can't be, e.g, cloned

What do you think?

Reply:
As discussed:

If the volume is created and not removed, the VM snapshot entry will remain in OK state, with a warning, but otherwise usable. Since the image is removed if we remove the new volume from the storage, the snapshot will be partial if we only  some of the volumes are left